include/OpenColorIO/OpenColorIO.h

```diff
@@ -634,6 +652,148 @@ class OCIOEXPORT Config
 
 extern OCIOEXPORT std::ostream& operator<< (std::ostream&, const Config&);
 
+//!cpp:class:: The File Rules are a set of filepath to color space mappings that are evaluated
+// from first to last. The first rule to match is what determines which color space is
+// returned. There are four types of rules available. Each rule type has a name key that may
+// be used by applications to refer to that rule. Name values must be unique. The other keys
+// depend on the rule type:
+//
+// - Basic Rule: This is the basic rule type that uses Unix glob style pattern matching and
+// is thus very easy to use. It contains the keys:
+//   * name: Name of the rule
+//   * colorspace: Color space name to be returned.
+//   * pattern: Glob pattern to be used for the main part of the name/path.
+//   * extension: Glob pattern to be used for the file extension. Note that if glob tokens
+//     are not used, the extension will be used in a non-case-sensitive way by default.
+// 
+// - Regex Rule: This is similar to the basic rule but allows additional capabilities for
+//   power-users. It contains the keys:
+//   * name: Name of the rule
+//   * colorspace: Color space name to be returned.
+//   * regex: Regular expression to be evaluated.
+// 
+// - OCIO v1 style Rule: This rule allows the use of the OCIO v1 style, where the string
+//   is searched for color space names from the config. This rule may occur 0 or 1 times
+//   in the list. The position in the list prioritizes it with respect to the other rules.
+//   It has the key:
+//   * name: Must be "ColorSpaceNamePathSearch".
+// 
+// - Default Rule: The file_rules must always end with this rule. If no prior rules match,
+//   this rule specifies the color space applications will use. It has the keys:
+//   * name: must be "Default".
+//   * colorspace : Color space name to be returned.
+//
+// Basic and Regex Rules can have custom string keys and associatated string values.
+//
+// Getters and setters are using the rule position, they will throw if the position is not
+// valid. If the rule at the specified position does not implement the requested property
+// getter will return NULL and setter will throw.
+//
+class FileRules
+{
+public:
+    //!cpp:function:: Creates FileRules for a Config. File rules will contain the default rule
+    // using the default role. The default rule can not be removed.
+    static FileRulesRcPtr Create();
+
+    //!cpp:function::
+    FileRulesRcPtr createEditableCopy() const;
+
+    //!cpp:function:: Does include default rule. Result will be at least 1.
+    size_t getNumEntries() const noexcept;
+
+    //!cpp:function:: Get name of the rule.
+    const char * getName(size_t ruleIndex) const;
+
+    //!cpp:function:: Setting pattern will erase regex.
```

Comment:
In my gut I feel like a `setRuleType` method (which would serialize to the config in a type attr) would be cleaner than relying on mutually exclusive attributes (clearing regex when glob is used and vice-versa). It would also prevent users from needing to manually enter either of the reserved rule type names, which could be set automatically when either of those types are used. You could then only have a `pattern` and `extension` attr (no `regex`, since it would be implied by the type and use the `pattern` slot). Just my two cents...

Reply:
Pattern/extension and regex do not use the same syntax/language, it would be strange to share the string, so changing the type would require the string to be cleared.
Maybe we could expose the reserved names in the API.

Reply:
Though if the type is mutually exclusive, the pattern would be clearly one or the other. I'm not stuck on that idea though. I think the thing that is a bit confusing with this class is that it can be many pre-defined things, each of which is fully determined by a subset of the attributes. Sometimes name drives behavior, other times regex, and other times pattern and extension. If we add more rules later it could balloon more. Willing to be wrong here of course, but maybe should be discussed more.

Reply:
Also, exposing reserved names is a great idea.

Reply:
Instead a function was added so that user does not have to specify the name of the rule.

Reply:
Thanks. Works well!